source/opt/iterator.h

```diff
@@ -0,0 +1,153 @@
+
+#include <memory>
+#include <vector>
+
+namespace spvtools {
+namespace ir {
+
+// An ad hoc iterator class for std::vector<std::unique_ptr<|ValueType|>>. The
+// purpose of this iterator class is to provide transparent access to those
+// std::unique_ptr managed elements in the vector, behaving like we are using
+// std::vector<|ValueType|>.
+template <typename ValueType, bool IsConst = false>
+class UptrVectorIterator {
+ public:
+  using pointer =
+      typename std::conditional<IsConst, const ValueType*, ValueType*>::type;
+  using reference =
+      typename std::conditional<IsConst, const ValueType&, ValueType&>::type;
+
+  // Type aliases. We need to apply constness properly if |IsConst| is true.
+  using Uptr = std::unique_ptr<ValueType>;
+  using UptrVector = typename std::conditional<IsConst, const std::vector<Uptr>,
+                                               std::vector<Uptr>>::type;
+  using UnderlyingIterator =
+      typename std::conditional<IsConst, typename UptrVector::const_iterator,
+                                typename UptrVector::iterator>::type;
+
+  // Creates a new iterator from the given |container| and its raw iterator
+  // |it|.
+  explicit UptrVectorIterator(UptrVector* container, UnderlyingIterator it)
+      : container_(container), iterator_(it) {}
+  UptrVectorIterator(const UptrVectorIterator&) = default;
+  UptrVectorIterator& operator=(const UptrVectorIterator&) = default;
+
+  inline UptrVectorIterator operator++();
+  inline UptrVectorIterator operator++(int);
+  inline UptrVectorIterator operator--();
+  inline UptrVectorIterator operator--(int);
+
+  reference operator*() const { return **iterator_; }
+  pointer operator->() { return (*iterator_).get(); }
+  reference operator[](ptrdiff_t index) { return **(iterator_ + index); }
+
+  inline bool operator==(const UptrVectorIterator& that) const;
+  inline bool operator!=(const UptrVectorIterator& that) const;
+
+  // Inserts the given |value| to the position pointed to by this iterator
+  // and returns an iterator to the newly iserted |value|.
+  // If the underlying vector changes capacity, all previous iterators will be
+  // invalidated. Otherwise, those previous iterators pointing to after the
+  // insertion point will be invalidated.
+  inline UptrVectorIterator InsertBefore(Uptr value);
+
+ private:
+  UptrVector* container_;        // The container we are manipulating.
+  UnderlyingIterator iterator_;  // The raw iterator from the container.
+};
+
+// Handly class for a (begin, end) iterator pair.
+template <typename IteratorType>
+class IteratorRange {
+ public:
+  IteratorRange(IteratorType b, IteratorType e) : begin_(b), end_(e) {}
+
+  IteratorType begin() const { return begin_; }
+  IteratorType end() const { return end_; }
+
+  bool empty() const { return begin_ == end_; }
+  size_t size() const { return end_ = begin_; }
+
+ private:
+  IteratorType begin_;
+  IteratorType end_;
+};
+
+template <typename VT, bool IC>
```

Comment:
I feel like this is more concise since we are not far away from the above full declaration.